// crypto-wallet-core/ts_build/src/validation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bch_1 = require("./bch");
var btc_1 = require("./btc");
var doge_1 = require("./doge");
var eth_1 = require("./eth");
var ltc_1 = require("./ltc");
var xrp_1 = require("./xrp");
var validation = {
    BTC: new btc_1.BtcValidation(),
    BCH: new bch_1.BchValidation(),
    ETH: new eth_1.EthValidation(),
    XRP: new xrp_1.XrpValidation(),
    DOGE: new doge_1.DogeValidation(),
    LTC: new ltc_1.LtcValidation()
};
var ValidationProxy = (function () {
    function ValidationProxy() {
    }
    ValidationProxy.prototype.get = function (chain) {
        var normalizedChain = chain.toUpperCase();
        return validation[normalizedChain];
    };
    ValidationProxy.prototype.validateAddress = function (chain, network, address) {
        return this.get(chain).validateAddress(network, address);
    };
    ValidationProxy.prototype.validateUri = function (chain, addressUri) {
        return this.get(chain).validateUri(addressUri);
    };
    return ValidationProxy;
}());
exports.ValidationProxy = ValidationProxy;
exports.default = new ValidationProxy();
