// crypto-wallet-core/ts_build/src/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var BitcoreLib = __importStar(require("bitcore-lib"));
exports.BitcoreLib = BitcoreLib;
var BitcoreLibCash = __importStar(require("bitcore-lib-cash"));
exports.BitcoreLibCash = BitcoreLibCash;
var BitcoreLibDoge = __importStar(require("bitcore-lib-doge"));
exports.BitcoreLibDoge = BitcoreLibDoge;
var BitcoreLibLtc = __importStar(require("bitcore-lib-ltc"));
exports.BitcoreLibLtc = BitcoreLibLtc;
var web3_1 = __importDefault(require("web3"));
exports.Web3 = web3_1.default;
var constants_1 = require("./constants");
exports.Constants = constants_1.Constants;
var derivation_1 = __importDefault(require("./derivation"));
exports.Deriver = derivation_1.default;
var transactions_1 = __importDefault(require("./transactions"));
exports.Transactions = transactions_1.default;
var validation_1 = __importDefault(require("./validation"));
exports.Validation = validation_1.default;
